spring-cloud-deployer-local/src/main/java/org/springframework/cloud/deployer/spi/local/LocalAppDeployer.java

```diff
@@ -206,13 +207,40 @@ public AppStatus status(String id) {
 		return builder.build();
 	}
 
+	/**
+	 * Shut down the {@link Process} backing the application {@link Instance}.
+	 * If the application exposes a {@code /shutdown} endpoint, that will be
+	 * invoked followed by a wait that will not exceed the number of milliseconds
+	 * indicated by {@link LocalDeployerProperties#shutdownTimeout}. If the
+	 * timeout period is exceeded (or if the {@code /shutdown} endpoint is not exposed),
+	 * the process will be shut down via {@link Process#destroy()}.
+	 *
+	 * @param instance the application instance to shut down
+	 */
 	private void shutdownAndWait(Instance instance) {
 		try {
-			restTemplate.postForObject(instance.url + "/shutdown", null, String.class);
-			instance.process.waitFor();
+			int timeout = getLocalDeployerProperties().getShutdownTimeout();
+			if (timeout > 0) {
+				ResponseEntity<String> response = restTemplate.postForEntity(
+						instance.url + "/shutdown", null, String.class);
+				if (response.getStatusCode().is2xxSuccessful()) {
+					long timeoutTimestamp = System.currentTimeMillis() + timeout;
+					while (isAlive(instance.process) && System.currentTimeMillis() < timeoutTimestamp) {
+						Thread.sleep(1000);
```

Comment:
If we're going to wait a second here anyways, I think it might be ok to have the timeout granularity (the `TimeUnit` enum value basically) be *seconds* rather than milliseconds. Otherwise, any value would be effectively rounded to the next second anyways. WDYT?